app/src/main/java/io/homeassistant/companion/android/nfc/NfcActivity.kt

```diff
@@ -0,0 +1,339 @@
+                }
+
+                // Fetch entities
+                integrationUseCase.getEntities().forEach {
+                    entities[it.entityId] = it
+                }
+            } catch (e: Exception) {
+                // Custom components can cause services to not load
+                // Display error text
+                widget_config_service_error.visibility = View.VISIBLE
+            }
+        }
+
+        widget_text_config_service.addTextChangedListener(serviceTextWatcher)
+
+        add_field_button.setOnClickListener(onAddFieldListener)
+        add_button.setOnClickListener(onClickListener)
+
+        dynamicFieldAdapter = WidgetDynamicFieldAdapter(services, entities, dynamicFields)
+        widget_config_fields_layout.adapter = dynamicFieldAdapter
+        widget_config_fields_layout.layoutManager = LinearLayoutManager(this)
+
+
+        write_tag_id.setOnClickListener {
+            nfcUrlToWrite = "https://www.home-assistant.io/nfc/${UUID.randomUUID()}"
+            Toast.makeText(applicationContext, R.string.nfc_write_tag_info, Toast.LENGTH_LONG)
+                .show()
+        }
+    }
+
+    override fun onResume() {
+        super.onResume()
+        mNfcAdapter?.let {
+            NFCUtil.enableNFCInForeground(it, this,javaClass)
+        }
+    }
+
+    override fun onPause() {
+        super.onPause()
+        mNfcAdapter?.let {
+            NFCUtil.disableNFCInForeground(it,this)
+        }
+    }
+
+    override fun onDestroy() {
+        mainScope.cancel()
+        super.onDestroy()
+    }
+
+
+    private fun String.utf8(): String = java.net.URLEncoder.encode(this, "UTF-8")
+    private var onClickListener = View.OnClickListener {
+        try {
+            val context = this@NfcActivity
+
+            // Analyze and send service and domain
+            val serviceText = context.widget_text_config_service.text.toString()
+            val domain = services[serviceText]?.domain ?: serviceText.split(".", limit = 2)[0]
+            val service = services[serviceText]?.service ?: serviceText.split(".", limit = 2)[1]
+
+            // Analyze and send service data
+            val serviceDataMap = HashMap<String, List<Any>>()
+            dynamicFields.forEach {
+                if (it.value != null) {
+                    serviceDataMap[it.field] = it.value!! as List<Any>
+                }
+            }
+
+            val params = serviceDataMap.map {(k, v) -> "${k.utf8()}=${v[0]}"}.joinToString("&")
+
+            nfcUrlToWrite =
+                "https://www.home-assistant.io/nfc/?url=homeassistant://call_service/$domain.$service?$params"
+            Log.d(TAG, "Generated url: $nfcUrlToWrite")
+            Toast.makeText(applicationContext, R.string.nfc_write_tag_info, Toast.LENGTH_LONG)
+                .show()
+        } catch (e: Exception) {
+            Log.e(TAG, "Issue configuring nfc tag", e)
+            Toast.makeText(applicationContext, R.string.widget_creation_error, Toast.LENGTH_LONG)
```

Comment:
Hope it's not a widget creation error! ;-)
Also we might want to do more than just a toast to let the user know something went wrong.

Reply:
Ahah oh no.. my bad, fixed it! Thanks for noticing!